### PR TITLE
[Arith] add SizeVar representing non-neg valued variable in a tensor shape

### PR DESCRIPTION
**Pull request by yzhliu**
To provide extra information for arith simplification.

This is an alternative approach for https://github.com/apache/incubator-tvm/pull/4486 

More background,

https://discuss.tvm.ai/t/discuss-embed-more-bound-information-into-var-or-expr

https://discuss.tvm.ai/t/significant-increase-in-the-amount-of-cuda-code-gen-after-migrating-indexdiv-mod-to-floordiv-mod

I haven't change `tvm.var` to `tvm.shape_var` in topi because @icemelon9 is in the middle of refactoring https://github.com/apache/incubator-tvm/pull/4644 . We can do it later.

@tqchen @icemelon9 Please review

### CODE REVIEW

**Comment by yzhliu** [comment]
@tqchen do you have any idea how shape_var fails to be in func arg after I rebase from upstream today?

**Comment by yzhliu** [comment]
@tqchen @icemelon9 CI's green. it's ready for review.

File: include/tvm/expr.h [hunk 1]
```diff
@@ -115,6 +115,43 @@ class Var : public PrimExpr {
...
 };
 
+class ShapeVar;
+/*!
+ * \brief A variable node represent a tensor shape size,
+ * whose value must be non-negative.
+ */
+class ShapeVarNode : public VarNode {
+ public:
+  static ShapeVar make(DataType dtype, std::string name_hint);
```
Comment (tqchen) [inline include/tvm/expr.h]: use constructor in ShapeVar, as per https://github.com/apache/incubator-tvm/issues/4648

File: python/tvm/build_module.py [hunk 2]
```diff
@@ -309,7 +313,7 @@ def get_binds(args, compact=False, binds=None):
                 arg_list.append(binds[x])
         elif isinstance(x, schedule.Buffer):
             arg_list.append(x)
-        elif isinstance(x, expr.Var):
```
Comment (tqchen) [inline python/tvm/build_module.py]: keep the original one, as we can make ShapeVar as subclass expr.Var

**Comment by tqchen** [comment]
cc @ZihengJiang @Hzfengsy would be great if you can help to take a look

**Comment by yzhliu** [comment]
passing var into the compute still works

File: src/arithmetic/bound_deducer.cc [hunk 3]
```diff
@@ -70,6 +70,40 @@ std::vector<const Object*> GetPath(PrimExpr target, PrimExpr expr) {
   return v.path_;
 }
 
+class BoundRemover : public ExprMutator {
+ public:
+  PrimExpr Remove(const PrimExpr& e) {
+    remove_bounded_ = true;
```
Comment (tqchen) [inline src/arithmetic/bound_deducer.cc]: Can you document its behavior? Is it still necessary, given bound info is no longer wrapped in the assert_

File: src/arithmetic/bound_deducer.cc [hunk 4]
```diff
@@ -297,6 +331,18 @@ void BoundDeducer::Transform() {
 void BoundDeducer::Deduce() {
   Init();
   if (!success_) return;
+
+  // Any variable appears in both expr and result,
```
Comment (tqchen) [inline src/arithmetic/bound_deducer.cc]: I do not understand the following comment. Why would n/4 be simplified to 0? if n>=0 the simplification rule does not stand, if n< 4, then it is a valid simplification

File: src/arithmetic/int_set.cc [hunk 5]
```diff
@@ -405,6 +405,23 @@ class IntervalSetEvaluator :
...
+      IntervalSet res = ToIntervalSet((*it).second);
+      if (res->min_value.same_as(var) &&
+          res->max_value.same_as(var)) {
+        return res;
+      }
+      // recursively evaluate mapped result
+      // in case the domain contains variables to be relaxed.
+      return Eval(res);
+    } else {
+      return IntervalSet(0, GetRef<ShapeVar>(op));
```
Comment (tqchen) [inline src/arithmetic/int_set.cc]: Shouldn't it be IntervalSet::SinglePoint(var) as in Var? This is a bit overly relaxed

File: src/arithmetic/int_set.cc [hunk 6]
```diff
@@ -519,6 +536,14 @@ class IntervalSetEvaluator :
     return set->min_value.same_as(value) && set->max_value.same_as(value);
   }
 
+  bool SelfBoundedVar(const IntervalSet& set,
+                      const PrimExpr& value) const {
```
Comment (tqchen) [inline src/arithmetic/int_set.cc]: need comment about the function name.

File: src/arithmetic/int_set.cc [hunk 7]
```diff
@@ -529,6 +554,22 @@ class IntervalSetEvaluator :
     return Combine<T>(analyzer_, a, b);
   }
 
+  template<typename T>
+  inline IntervalSet VisitDivExpr_(const T* op) {
+    IntervalSet a = this->Eval(op->a);
+    IntervalSet b = this->Eval(op->b);
+    if ((MatchPoint(a, op->a) && (MatchPoint(b, op->b) || SelfBoundedVar(b, op->b)))
+          || (SelfBoundedVar(a, op->a) && SelfBoundedVar(b, op->b))) {
+      // e.g.,
```
Comment (tqchen) [inline src/arithmetic/int_set.cc]: This part is a bit confusing. If the divisor is already a set(which means it is relaxed), we should not return the original var.

File: src/arithmetic/bound_deducer.cc (hunk 4 above)
Comment (yzhliu) [reply src/arithmetic/bound_deducer.cc]: for `>=`, current logic is to first get the lower bound of the left side then deduce for `i`.

File: src/arithmetic/int_set.cc (hunk 7 above)
Comment (yzhliu) [reply src/arithmetic/int_set.cc]: but in `Everything<Div>` such case evaluates to `IntervalSet::Everything()`. 
e.g., `4 / tvm.var() => 4 / tvm.var()`, while `4 / tvm.shape_var() => (-inf, +inf)`

File: src/arithmetic/int_set.cc (hunk 5 above)
Comment (yzhliu) [reply src/arithmetic/int_set.cc]: fair. I'll change and verify.

File: src/arithmetic/bound_deducer.cc (hunk 4 above)
Comment (tqchen) [reply src/arithmetic/bound_deducer.cc]: I see, perhaps we can add more explanation then. 

Also, would it results in an incorrect bound? or just a more strict bound(which still makes the condition holds)

For a better approach. My feeling is that we should first have a rewriting that tries to move vars into one side

File: src/arithmetic/int_set.cc (hunk 7 above)
Comment (tqchen) [reply src/arithmetic/int_set.cc]: i think this can be fixed by having shape_var return single_pt

File: src/arithmetic/bound_deducer.cc (hunk 4 above)
Comment (yzhliu) [reply src/arithmetic/bound_deducer.cc]: I believe it is just a more strict bound.
I agree, let me have a try to see if it is feasible.

File: src/arithmetic/int_set.cc (hunk 7 above)
Comment (yzhliu) [reply src/arithmetic/int_set.cc]: yeah... you're right.

File: src/arithmetic/bound_deducer.cc (hunk 4 above)
Comment (yzhliu) [reply src/arithmetic/bound_deducer.cc]: as it is also using `IntervalSetEvaluator`, this is also not required after having shape_var return single_pt

**Comment by yzhliu** [comment]
@tqchen updated per comments. do you think there could be a better name than `shape_var`? @icemelon9 reminded there is `ShapeVar` in relay, which represents an n-dim tuple.

**Comment by tqchen** [comment]
I don't have a better idea for name. Indeed shape could indicate a tuple rather than an integer. 

We could potentially rename the relay's shape template variable for clarity as it is really type var. But we could think about a better name.

File: include/tvm/expr.h [hunk 8]
```diff
@@ -115,6 +117,45 @@ class Var : public PrimExpr {
...
+
+  static constexpr const char* _type_key = "ShapeVar";
+  TVM_DECLARE_FINAL_OBJECT_INFO(ShapeVarNode, VarNode);
+};
+
+/*! \brief a named variable represents a tensor shape size */
+class ShapeVar : public Var {
+ public:
+  explicit ShapeVar(ObjectPtr<Object> n) : Var(n) {}
+  TVM_DLL explicit ShapeVar(std::string name_hint = "s",
```
Comment (tqchen) [inline include/tvm/expr.h]: Document the constructor

**Comment by yzhliu** [comment]
How about `TIndexVar` stands for tensor index variable?

**Comment by tqchen** [comment]
We can ask other's thoughts about it(perhaps list a few candidates and send an rfc?). TIndex seems to be a reasonable name, although it is a bit ambiguous because people need to guess what does T mean).

**Comment by yzhliu** [comment]
https://discuss.tvm.ai/t/rfc-name-for-non-negative-variable

**Comment by yzhliu** [comment]
@tqchen I have address the comments and changed the name

File: python/tvm/api.py [hunk 9]
```diff
@@ -192,6 +192,25 @@ def var(name="tindex", dtype=int32):
     return _api_internal._Var(name, dtype)
 
 
+def size_var(name="tindex", dtype=int32):
```
Comment (tqchen) [inline python/tvm/api.py]: is tindex a good default? size?

**Comment by tqchen** [comment]
cc @yzhliu feel free to merge after fixing the nits. BTW, it also reminds me if we would like to bring IterVar as a sub-class of Var, right now it takes Var as a member.

**Comment by yzhliu** [comment]
Thanks @tqchen I'll take a look into the IndexVar and send another pr

**Pull request merged**